### PR TITLE
Form Builder

### PR DESCRIPTION
**Pull request by aptkingston**
## Description
This PR adds the functionality to create custom forms. Form fields are individual components in the builder and can be add, moved, styled and deleted as desired. Forms are built of 3 components - forms, field groups and fields.

forms

### Forms
Forms are the top level component. All fields must be placed within a form, and placing a field outside a form will helpfully inform the user of this. Forms can be quickly styled via the "Theme" and "Size" settings. Autoscreens which contain forms are pre-styled to support theming - so even the breadcrumbs and form titles will respond to form theme changes. Forms can contain any content - not just field groups and fields - so you can easily nest any other components alongside your form components. Field groups are the easiest way to achieve this neatly. 

Forms have a schema setting which is optional. The schema for a form can be a table, view, relationship, query or "custom". A custom schema means that there is no backing datasource schema.
Choosing a form schema has the following advantages:
- When adding fields, field names can be chosen from a dropdown rather than being typed
- Field validation is automatically added to match the schema
- If you want to use the "Save Row" action using your form data, you need to have a schema to provide the associated table ID
- Fields will be able to automatically prevent you entering an illegal field name for a certain data type
- Form components can be automatically generated to match the schema

Forms provide a new button action to validate the form. This action will trigger the form validation, show any errors, and stop executing any further actions if the form is invalid. The recommended way to use this is for your first button action to be validating the form, and the second action to be performing your task - for example saving a row. This ensures only valid forms will be submitted.

### Field Groups
Field groups are container components for groups of fields. Field groups are optional - you can add fields directly to a form component - but if you want to change the field label position (for example, neatly aligning them to left of the fields), then you'll need to group the components in a field group. Field groups should **only** contain fields - they should not contain any non-field components. This is because they rely on `display: table` to align labels, and this layout will not work nicely with components that aren't fields.

Field groups have only 1 setting, the label position. Changing this will align the labels of all fields in the field group together.

When a field group is selected, a button named "Reset Fields" will be displayed in the settings panel. This button will reset the content of the field group to match the schema of the form, replacing all existing fields. A confirmation modal ensures this can't happen by accident. This button is deliberately set on the field group component rather than the form - because forms can contain other data that you may not want to delete. By nesting fields in a field group, you can design a rich, custom form and only reset the field group component whenever you change your backend schema.

Here's a video showing this functionality.
Untitled_ Feb 9, 2021 3_01 PM (1)

### Fields
Each field component represents a single field in the form. All fields have a `field` and a `label` setting. The `field` is the name of the field, or property of the schema. This can be free typed or, if a schema exists on the form, selected from a list. There are unique components for every type of data, and if your form has a schema then you'll only be able to choose field names which fit the particular type of component that is selected. The `label` setting is the label that will be displayed beside or above your field. Most fields also have a `placeholder` setting to control the text displayed when the field is empty.

Whether your form has a schema or not, you can always add custom fields that don't exist in the schema by free typing the name of the field. Data bindings will always reflect the names of all the fields in your form - so even if you don't have a schema on your form, you'll be able to bind to any of your fields.

## Adding Relational Data
Another feature of this PR is the ability to easily add related data. This could be adding data related to the logged in user, or adding data to the row detail page you're looking at. This is achieved by binding to an ID for a relationship field in something like a "Save Row" action. This didn't used to be possible as the server enforced relationships being arrays - but thanks to @mike12345567, the server will now coerce a string value into a single array value.

Here's a video showing this functionality - note that there's only 1 field being entered, yet it creates the relationship.
Untitled_ Feb 9, 2021 2_52 PM

Here's what the builder settings look like to achieve this functionality.
image

## Data Hot Reloading
One other feature which has been snuck into this PR is automatic hot reloading of data. For example - if you add a new row and you also happen to have a repeater showing those rows - you'll now instantly see the new record. If you update a row, you'll instantly see the changes reflected. This works for all data sources - tables, views, relationships and external data sources. The best part is that this is zero config. You don't need to change anything in your builder, it just works.

Here's a video showing this functionality - note how everything updates in real time.
Untitled_ Feb 9, 2021 2_56 PM

### CODE REVIEW

**Comment by joebudi** [comment]
This one PR lays the foundation for many varieties of apps that can be developed on Budibase. Incredible work. This is truly a USP for Budibase. Well done @aptkingston

**Review by shogunpurple — APPROVED** [review]
Incredible job @aptkingston - this will be a truly core feature of budibase going forward and set the precedent for how bb apps interact with data.

File: packages/builder/yarn.lock [hunk 1]
```diff
@@ -842,10 +842,17 @@
     lodash "^4.17.19"
     to-fast-properties "^2.0.0"
 
+<<<<<<< HEAD
```
Comment (shogunpurple) [inline packages/builder/yarn.lock:845]: Little merge conflict here

Comment (aptkingston) [reply packages/builder/yarn.lock:845]: I knew it! I actually checked for this and Webstorm was like nah there's no errors in this file. Thanks for spotting.

**Comment by mike12345567** [comment]
I've been so excited for this for so long, its truly incredible work 👏 👏 👏

**Pull request merged**